File: numass-control/vac/src/main/java/inr/numass/control/readvac/MeradatVacDevice.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package inr.numass.control.readvac;

import hep.dataforge.context.Context;
import hep.dataforge.control.devices.Device;
import hep.dataforge.control.devices.PortSensor;
import hep.dataforge.control.measurements.Measurement;
import hep.dataforge.control.measurements.SimpleMeasurement;
import hep.dataforge.control.ports.PortHandler;
import hep.dataforge.description.ValueDef;
import hep.dataforge.exceptions.ControlException;
import hep.dataforge.meta.Meta;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static hep.dataforge.values.ValueType.NUMBER;

/**
 * @author Alexander Nozik
 */
@ValueDef(name = "address", type = {NUMBER}, def = "1", info = "A modbus address")
public class MeradatVacDevice extends PortSensor<Double> {
    private static final String REQUEST = "0300000002";

    public MeradatVacDevice() {
    }

    public MeradatVacDevice(Context context, Meta meta) {
        setContext(context);
        setMeta(meta);
    }

    @Override
    protected PortHandler buildHandler(String portName) throws ControlException {
        PortHandler newHandler = super.buildHandler(portName);
        newHandler.setDelimiter("\r\n");
        return newHandler;
    }

    @Override
    protected Measurement<Double> createMeasurement() {
        return new MeradatMeasurement();
    }

    @Override
    public String getType() {
        return meta().getString("type", "Vit vacuumeter");
    }

    public static String calculateLRC(String inputString) {
        /*
         * String is Hex String, need to convert in ASCII.
         */
        byte[] bytes = new BigInteger(inputString, 16).toByteArray();
        int checksum = 0;
        for (byte aByte : bytes) {
            checksum += aByte;
        }
        String val = Integer.toHexString(-checksum);
        val = val.substring(val.length() - 2).toUpperCase();
        if (val.length() < 2) {
            val = "0" + val;
        }

        return val;
    }


    private class MeradatMeasurement extends SimpleMeasurement<Double> {

        private final String query; // ":010300000002FA\r\n";
        private final Pattern response;
        private final String base;

        public MeradatMeasurement() {
            base = String.format(":%02d", meta().getInt("address", 1));
            String dataStr = base.substring(1) + REQUEST;
            query = base + REQUEST + calculateLRC(dataStr) + "\r\n";
            response = Pattern.compile(base + "0304(\\w{4})(\\w{4})..\r\n");
        }

        @Override
        protected synchronized Double doMeasure() throws Exception {

            String answer = sendAndWait(query, timeout(), phrase -> phrase.startsWith(base));

            if (answer.isEmpty()) {
                this.updateMessage("No signal");
                updateState(CONNECTED_STATE, false);
                return null;
            } else {
                Matcher match = response.matcher(answer);

                if (match.matches()) {
                    double base = (double) (Integer.parseInt(match.group(1), 16)) / 10d;
                    int exp = Integer.parseInt(match.group(2), 16);
                    if (exp > 32766) {
                        exp = exp - 65536;
                    }
                    BigDecimal res = BigDecimal.valueOf(base * Math.pow(10, exp));
                    res = res.setScale(4, RoundingMode.CEILING);
                    this.updateMessage("OK");
                    updateState(CONNECTED_STATE, true);
                    return res.doubleValue();
                } else {
                    this.updateMessage("Wrong answer: " + answer);
                    updateState(CONNECTED_STATE, false);
                    return null;
                }
            }
        }

        @Override
        public Device getDevice() {
            return MeradatVacDevice.this;
        }
    }

}

File: numass-control/vac/src/main/kotlin/inr/numass/control/readvac/MeradatVacDevice.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package inr.numass.control.readvac

import hep.dataforge.context.Context
import hep.dataforge.control.devices.PortSensor
import hep.dataforge.control.ports.GenericPortController
import hep.dataforge.control.ports.Port
import hep.dataforge.control.ports.PortFactory
import hep.dataforge.description.ValueDef
import hep.dataforge.meta.Meta
import hep.dataforge.states.StateDef
import hep.dataforge.states.valueState
import hep.dataforge.values.ValueType.NUMBER
import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode
import java.util.regex.Pattern

/**
 * @author Alexander Nozik
 */
@StateDef(value = ValueDef(key = "address", type = [NUMBER], def = "1", info = "A modbus address"), writable = true)
class MeradatVacDevice(context: Context, meta: Meta) : PortSensor(context, meta) {

    var address by valueState("address").intDelegate

    override fun buildConnection(meta: Meta): GenericPortController {
        val port: Port = PortFactory.build(meta)
        logger.info("Connecting to port {}", port.name)

        return GenericPortController(context, port) { it.endsWith("\r\n") }
    }

    override val type: String
        get() {
            return meta.getString("type", "numass.vac.vit")
        }

    override fun startMeasurement(oldMeta: Meta?, newMeta: Meta) {
        measurement{
            val requestBase: String = String.format(":%02d", address)
            val dataStr = requestBase.substring(1) + REQUEST
            val query = requestBase + REQUEST + calculateLRC(dataStr) + "\r\n" // ":010300000002FA\r\n";
            val response: Pattern = Pattern.compile(requestBase + "0304(\\w{4})(\\w{4})..\r\n")

            val answer = sendAndWait(query) { phrase -> phrase.startsWith(requestBase) }

            if (answer.isEmpty()) {
                updateState(PortSensor.CONNECTED_STATE, false)
                notifyError("No signal")
            } else {
                val match = response.matcher(answer)

                if (match.matches()) {
                    val base = Integer.parseInt(match.group(1), 16).toDouble() / 10.0
                    var exp = Integer.parseInt(match.group(2), 16)
                    if (exp > 32766) {
                        exp -= 65536
                    }
                    var res = BigDecimal.valueOf(base * Math.pow(10.0, exp.toDouble()))
                    res = res.setScale(4, RoundingMode.CEILING)
                    updateState(PortSensor.CONNECTED_STATE, true)
                    notifyResult(res)
                } else {
                    updateState(PortSensor.CONNECTED_STATE, false)
                    notifyError("Wrong answer: $answer")
                }
            }
        }
    }

    companion object {
        private const val REQUEST = "0300000002"

        fun calculateLRC(inputString: String): String {
            /*
         * String is Hex String, need to convert in ASCII.
         */
            val bytes = BigInteger(inputString, 16).toByteArray()
            val checksum = bytes.sumOf { it.toInt() }
            var value = Integer.toHexString(-checksum)
            value = value.substring(value.length - 2).uppercase()
            if (value.length < 2) {
                value = "0$value"
            }

            return value
        }
    }
}
